Add a script that finds deprecated items.

drwells opened the pull request:
Closes #3965. This makes it easy to find everything in the `include` directory that has been deprecated. The output format is compatible with emacs' compilation mode in case people want to work with things there.

davydden commented on contrib/utilities/print_deprecated.py:
```diff
@@ -0,0 +1,103 @@
+                if 0 <= self.line_n - line_n <= n_context_lines:
+                    self.context_lines.append(file_line)
+
+        git_log_output = subprocess.check_output(["git", "blame", "-p",
```
i wonder if one could get from git commit the closest tag in history backwards? Say if something is deprecated right now, we could get to `v8.5.0` -- that's all we need, probably?

davydden replied:
one can do `git describe --abbrev=0`, but we need to be consistent with tags. Currently the closest tag is `rebase_here` 😮

davydden replied:
oh, here's a good way to do it:
```
git describe --tags $(git rev-list --tags --max-count=1)
```

davydden replied:
there is also an option
```
--contains            find the tag that comes after the commit
```

drwells commented:
@davydden Where does the relevant commit hash go in that `git describe` command? I can't make it print out the nearest tag for me.

davydden commented:
> Where does the relevant commit hash go in that git describe command?

at the end, i.e. `git describe --abbrev=0 <hash>`

drwells commented:
I am still missing something. If I do:

```
[drwells@learch dealii-dev]$ python ./contrib/utilities/print_deprecated.py | head -n 30

[...]

Date: 2015-04-08 15:56:25
Message: Deprecate the n_rows() and n_cols() member functions. The SparseMatrixEZ class
File: ./include/deal.II/lac/block_sparse_matrix_ez.h:168:
Line: size_type n_cols () const DEAL_II_DEPRECATED;
Hash: 1f1c471914abe70db2dafe91178fb7e174ceffe7

[drwells@learch dealii-dev]$ git describe --abbrev 1f1c471914abe
rebase_here-1680-g1f1c471914
```

which is not the tag, nor can I figure out how to make that print 8.4.0 (or whatever version is closest).

davydden commented:
@drwells indeed, i was not able to make `describe` work, BUT i think we are now one step closer with 
```
$ git tag --contains 1f1c471914abe70db2dafe91178fb7e174ceffe7 -l "v*"
v8.3.0
v8.3.0-rc1
v8.3.0-rc2
v8.3.rc1
v8.3.rc2
v8.4.0
v8.4.0-rc2
v8.4.0-rc3
v8.4.1
v8.4.2
v8.4.2-rc1
v8.5.0
v8.5.0-rc1
v8.5.0-rc2
```
then, you can take the last one which does not have `-rc`.

davydden commented:
or one can do
```
$ git tag --contains 1f1c471914abe70db2dafe91178fb7e174ceffe7 -l 'v[0-9].[0-9].[0-9]'
```

davydden commented:
@drwells here's another working way (don't ask my how/why it works):
```
git describe $(git rev-list --tags --max-count=1 1f1c471914abe70db2dafe91178fb7e174ceffe7)
```

davydden commented:
i think `tag --contains` is what we want. I am still not sure how `describe` works and what it actually does.

tamiko commented:
@davydden `describe` tries to describe a given commit by the closest tag (backwards in history) and how many commits come on top of it. With our tags exclusively on release branches we simply cannot `describe` any regular commit on master :-)

davydden commented:
@tamiko what about `tag --contains`? That seems to do what we want

davydden commented:
for the @drwells 's example commit, dated `2015-04-08`, `8.3.0` (2 Aug 2015) is indeed the first release which have it.

drwells commented:
It looks like `git tag --contains` is the way to go (as long as one has all relevant release branches in their local clone). I updated the script to print out the hash and release information and I added a bunch of documentation (and fixed some `pylint` complaints).

davydden commented on contrib/utilities/print_deprecated.py:
```diff
@@ -0,0 +1,138 @@
+        self.file_name = line_contents[0]
+        self.line_n = int(line_contents[1])
+
+        n_context_lines = 6
```
could you please document what this is and why it's `6`?

davydden replied:
ah, i see from below... maybe a quick comment right before this:
```
# open the file and get the context around the provided line
```

drwells commented:
@davydden I originally stored some extra lines in the file with the intent of printing out the deprecation line and the lines before it, but I never actually used this. I just got rid of it and cleaned up the parsing logic.
